File: src/phorge/api/endpoints/databases.py
"""Database and database user API endpoints."""

from __future__ import annotations

from typing import Any

from phorge.api.client import ForgeClient
from phorge.api.models import Database, DatabaseUser


class DatabasesAPI:
    def __init__(self, client: ForgeClient) -> None:
        self._client = client

    async def list(self, server_id: int) -> list[Database]:
        data = await self._client.get(f"/servers/{server_id}/databases")
        return [Database(**d) for d in data.get("databases", [])]

    async def get(self, server_id: int, database_id: int) -> Database:
        data = await self._client.get(f"/servers/{server_id}/databases/{database_id}")
        return Database(**data["database"])

    async def create(self, server_id: int, name: str, user: str | None = None, password: str | None = None) -> Database:
        payload: dict[str, Any] = {"name": name}
        if user:
            payload["user"] = user
        if password:
            payload["password"] = password
        data = await self._client.post(f"/servers/{server_id}/databases", json=payload)
        return Database(**data["database"])

    async def delete(self, server_id: int, database_id: int) -> None:
        await self._client.delete(f"/servers/{server_id}/databases/{database_id}")

    async def sync(self, server_id: int) -> None:
        await self._client.post(f"/servers/{server_id}/databases/sync")


class DatabaseUsersAPI:
    def __init__(self, client: ForgeClient) -> None:
        self._client = client

    async def list(self, server_id: int) -> list[DatabaseUser]:
        data = await self._client.get(f"/servers/{server_id}/database-users")
        return [DatabaseUser(**u) for u in data.get("users", [])]

    async def get(self, server_id: int, user_id: int) -> DatabaseUser:
        data = await self._client.get(f"/servers/{server_id}/database-users/{user_id}")
        return DatabaseUser(**data["user"])

    async def create(self, server_id: int, name: str, password: str, databases: list[int] | None = None) -> DatabaseUser:
        payload: dict[str, Any] = {"name": name, "password": password}
        if databases:
            payload["databases"] = databases
        data = await self._client.post(f"/servers/{server_id}/database-users", json=payload)
        return DatabaseUser(**data["user"])

    async def update(self, server_id: int, user_id: int, databases: list[int]) -> DatabaseUser:
        data = await self._client.put(
            f"/servers/{server_id}/database-users/{user_id}",
            json={"databases": databases},
        )
        return DatabaseUser(**data["user"])

    async def delete(self, server_id: int, user_id: int) -> None:
        await self._client.delete(f"/servers/{server_id}/database-users/{user_id}")

File: internal/forge/databases.go
package forge

import (
	"context"
	"net/http"
)

// List returns all databases on a server.
func (s *DatabasesService) List(ctx context.Context, serverID int64) ([]Database, error) {
	return listResources[Database](s.client, ctx, s.client.orgPath("/servers/%d/database/schemas", serverID))
}

// Get returns a single database by ID.
func (s *DatabasesService) Get(ctx context.Context, serverID, dbID int64) (*Database, error) {
	return getResource[Database](s.client, ctx, s.client.orgPath("/servers/%d/database/schemas/%d", serverID, dbID))
}

// Create creates a new database on a server.
// The user and password parameters are optional; pass nil to omit them.
func (s *DatabasesService) Create(ctx context.Context, serverID int64, name string, user, password *string) (*Database, error) {
	body := map[string]any{"name": name}
	if user != nil {
		body["user"] = *user
	}
	if password != nil {
		body["password"] = *password
	}
	return createResource[Database](s.client, ctx, s.client.orgPath("/servers/%d/database/schemas", serverID), body)
}

// Delete removes a database from a server.
func (s *DatabasesService) Delete(ctx context.Context, serverID, dbID int64) error {
	return s.client.do(ctx, http.MethodDelete, s.client.orgPath("/servers/%d/database/schemas/%d", serverID, dbID), nil, nil)
}

// Sync triggers a database sync on the server.
func (s *DatabasesService) Sync(ctx context.Context, serverID int64) error {
	return s.client.do(ctx, http.MethodPost, s.client.orgPath("/servers/%d/database/schemas/synchronizations", serverID), nil, nil)
}

// ListUsers returns all database users on a server.
func (s *DatabasesService) ListUsers(ctx context.Context, serverID int64) ([]DatabaseUser, error) {
	return listResources[DatabaseUser](s.client, ctx, s.client.orgPath("/servers/%d/database/users", serverID))
}

// GetUser returns a single database user by ID.
func (s *DatabasesService) GetUser(ctx context.Context, serverID, userID int64) (*DatabaseUser, error) {
	return getResource[DatabaseUser](s.client, ctx, s.client.orgPath("/servers/%d/database/users/%d", serverID, userID))
}

// CreateUser creates a new database user on a server.
func (s *DatabasesService) CreateUser(ctx context.Context, serverID int64, name, password string, databases []int64) (*DatabaseUser, error) {
	body := map[string]any{
		"name":     name,
		"password": password,
	}
	if databases != nil {
		body["databases"] = databases
	}
	return createResource[DatabaseUser](s.client, ctx, s.client.orgPath("/servers/%d/database/users", serverID), body)
}

// UpdateUser updates the database access for a database user.
func (s *DatabasesService) UpdateUser(ctx context.Context, serverID, userID int64, databases []int64) (*DatabaseUser, error) {
	path := s.client.orgPath("/servers/%d/database/users/%d", serverID, userID)
	body := map[string]any{"databases": databases}
	return updateResource[DatabaseUser](s.client, ctx, path, body)
}

// DeleteUser removes a database user from a server.
func (s *DatabasesService) DeleteUser(ctx context.Context, serverID, userID int64) error {
	return s.client.do(ctx, http.MethodDelete, s.client.orgPath("/servers/%d/database/users/%d", serverID, userID), nil, nil)
}
